### PR TITLE
PL Launch 2023 - Show default homepage hero banner to int'l audiences

### PR DESCRIPTION
**Pull request by kelbyhawn**
Show the default homepage hero banner to international audiences and the PL superhero banner to US only. 

**Asana task:** https://app.asana.com/0/0/1204402039772899

**Related PR:**
- https://github.com/code-dot-org/code-dot-org/pull/49588

### Testing story

[from Meg] The `homepage.rb` file currently has no tests. To convince myself that the `showInternationally` flag was working as expected, I used the announcements [test file](https://github.com/code-dot-org/code-dot-org/blob/0cc37cccdef003bc40c7c42b2eca3415436c9cb5/dashboard/test/lib/announcements_test.rb) for inspiration.

I had three attempted testing strategies and landed on the last one:
1) I was originally going to stub out the @@announcements_data variable using `Homepage.class_variable_set(:@@announcements_data, {...})`. But that approach meant that the vast majority of the test file was hard-to-parse JSONs, or hard-to-read updates to a single JSON.
2) Then, I was going to use a separate file as the announcements test file does. But we're needing to test banners with different `showInternationally` flags, and our current banner configuration returns the very first banner it finds:https://github.com/code-dot-org/code-dot-org/blob/0cc37cccdef003bc40c7c42b2eca3415436c9cb5/pegasus/src/homepage.rb#L34-L35
3) I ended up creating a separate file for each of the `showInternationally` flag configurations –– `true`, `false`, and no flag. I don't love that there is quite a bit of repeated code for each JSON, but I do think it is more readable than the approach in (1).

### Links

[from Meg] Asked Platform about best practices for querying location here https://codedotorg.slack.com/archives/CFTFD6BPV/p1681909723548929?thread_ts=1680634885.139079&cid=CFTFD6BPV

### CODE REVIEW

**Review by megcrenshaw — COMMENTED** [review]
I'm confused about the product requirement –– do we want to show the page for English users or for users inside the US? It looks like this PR will show the banner for any English user, including English users in, say, India or Australia.

**Comment by dmcavoy** [comment]
Good question Meg. The goal is to show the PL banner just to United States users. Is it possible to do that instead of relying on language to try to capture that?

**Comment by kelbyhawn** [comment]
> Good question Meg. The goal is to show the PL banner just to United States users. Is it possible to do that instead of relying on language to try to capture that?

I don't think we have the functionality for the homepage.json banners to show based off of geolocation (yet?) — I'll look into it.

**Comment by kelbyhawn** [comment]
For more context — this is what we did last year for the CSA launch hero banner that had similar audience requirements (Brendan helped me pair on this):
- https://github.com/code-dot-org/code-dot-org/pull/46406 ([file ref](https://github.com/code-dot-org/code-dot-org/pull/46406/files#diff-dd57fd05684ec1794d1ebd6ff8af9d3dc075ec811ab65bbe440cd8faa30a15a0))

**Comment by megcrenshaw** [comment]
Looks like this is where we set the language config:
https://github.com/code-dot-org/code-dot-org/blob/062d9efd4bf58238f1112ed4cfc9598a6fcc1191/pegasus/src/homepage.rb#L31-L32

I believe we could use a similar pattern for country, but it's not currently configured. Depending on how high priority this is, one of us could implement this functionality for, Kelby, you to use. (Kelby, I'm sure you could get this working too, but I'm guessing learning Ruby wasn't part of the job description 😅 .)

**Comment by dmcavoy** [comment]
@megcrenshaw thank you for looking into this. This is high priority as we are sending lots of traffic to pages that don't make any sense for them. Could you send some time today or tomorrow helping Kelby to get this working?

**Comment by kelbyhawn** [comment]
> Looks like this is where we set the language config:
> 
> https://github.com/code-dot-org/code-dot-org/blob/062d9efd4bf58238f1112ed4cfc9598a6fcc1191/pegasus/src/homepage.rb#L31-L32
> 
> I believe we could use a similar pattern for country, but it's not currently configured. Depending on how high priority this is, one of us could implement this functionality for, Kelby, you to use. (Kelby, I'm sure you could get this working too, but I'm guessing learning Ruby wasn't part of the job description 😅 .)

I just tried that in [this commit](https://github.com/code-dot-org/code-dot-org/pull/51356/commits/95cd97d69bd70c1e3aa720cf8159890154e4d647) but it wasn't working 😅 maybe I missed something though?

**Comment by megcrenshaw** [comment]
@dmcavoy Yep, on it!
@kelbyhawn Ah, brilliant! I missed that. You're way ahead of me. The `request` object doesn't include the country, but we can use this pattern:
```
location = Geocoder.search(user.current_sign_in_ip).first
country = location&.country_code&.to_s&.downcase
```

Want to try the following?
```
# If the banner has an array of countries, then the current country must be one of them.
location = Geocoder.search(user.current_sign_in_ip).first
country = location&.country_code&.to_s&.downcase
next if banner["countries"] && !banner["countries"].include?(country)
```

Then try `us` (lowercase) in the JSON. Or you can do `&.upcase` in the ruby file (instead of `&.downcase`) with capitalized `US` in the JSON –– whatever feels more intuitive to you.

If that doesn't work, let me know and we can either pair on it or I can take a deeper look solo. Thanks, Kelby, for starting on this!

File: pegasus/src/homepage.rb
Comment (kelbyhawn) [inline pegasus/src/homepage.rb:38]: @megcrenshaw pinging to remind you to set up tests for this 🙌

**Comment by megcrenshaw** [comment]
Follow-up ticket to add unit tests here https://codedotorg.atlassian.net/browse/ACQ-525

File: pegasus/src/homepage.rb
Comment (davidsbailey) [inline pegasus/src/homepage.rb]: questions:
1. how often is request.ip set? is it as reliable as user.current_sign_in_ip?
2. does this codepath work (no errors) when request.ip is nil?
3. in the case where request.ip is nil, are we ok with showing only outside us?
4. I'm not sure about taking away the option to display banners globally. isn't this something we'll want to do for HOC?

Comment (megcrenshaw) [reply pegasus/src/homepage.rb]: Good questions.
1-2) We can't use current_sign_in_ip because the user may not be signed in and we're on pegasus. But I don't know how often request.ip is set.
I went to this thread https://codedotorg.slack.com/archives/C0T0PNTM3/p1600671794018400 and it looks like the most established pattern is to be using `request.country` instead of `request.location.country_code`. @kelbyhawn thoughts about trying the following instead?
```
# remove location line
in_us = request.country.to_s.casecmp?('us')
next unless banner["showInternationally"] ^ in_us
```
3) @dmcavoy  do you know preferred behavior when we don't know the user's country?
4) There are three options with the `showInternationally` flag:
- `false` –– only show in the US
- `true` –– only show outside of the US
- no flag –– show everywhere

Comment (davidsbailey) [reply pegasus/src/homepage.rb]: > 4. There are three options with the `showInternationally` flag:
> 
> * `false` –– only show in the US
> * `true` –– only show outside of the US
> * no flag –– show everywhere

oh interesting! how is "show everywhere" implemented? the behavior I'm seeing makes me think that "no flag" will behave like `false`:
```
irb(main):001:0> nil ^ false
=> false
irb(main):002:0> nil ^ true
=> true
```

Comment (dmcavoy) [reply pegasus/src/homepage.rb]: If we don't know a users country I think it would be good to default to the US banner for now as thats our primary audience

Comment (megcrenshaw) [reply pegasus/src/homepage.rb]: Oh yikes, Dave! Good call –– we tested this manually, but I think didn't cover all the cases.

@kelbyhawn I think either (a) we pair again to get this working, or (b) I pull down the branch and make the commit myself. Do you have a preference?

Comment (davidsbailey) [reply pegasus/src/homepage.rb]: @megcrenshaw , given the need for thorough validation on our side, I'd encourage you to pull down the branch and make any changes needed, rather than going through Kelby for each update.

Comment (megcrenshaw) [reply pegasus/src/homepage.rb]: I might go ahead and add unit tests for this part too –– it will help me believe it's working

Comment (kelbyhawn) [reply pegasus/src/homepage.rb]: That works for me @megcrenshaw, and thanks @davidsbailey for reviewing this!

**Comment by megcrenshaw** [comment]
Just updated this file with refactored logic and some tests –– @breville and @davidsbailey would love for y'all to take a look when you're able! @kelbyhawn tagging you just to give you an update on progress.

File: pegasus/src/homepage.rb
Comment (kelbyhawn) [inline pegasus/src/homepage.rb]: Not blocking, very minor typo: `an` should be `a`.

Comment (megcrenshaw) [reply pegasus/src/homepage.rb]: Thanks!

File: dashboard/test/lib/homepage/homepage_test.rb
Comment (davidsbailey) [inline dashboard/test/lib/homepage/homepage_test.rb]: these tests are awesome, thanks for adding these!

I feel bad even mentioning this, but as part of the effort to separate dashboard from pegasus, we are trying to avoid cross-project dependencies like this, and so in the future any tests on pegasus pages should go in pegasus/test. this is the ideal dependency graph:

Screenshot 2023-04-19 at 9 29 26 AM

 this is my bad for not setting up linting rules preventing this kind of cross-project dependency, which I have been meaning to do for some time now. so for now I'm just going to ask that any additional tests for pegasus code be located in pegasus/test. 🙏

File: pegasus/src/homepage.rb
Comment (davidsbailey) [inline pegasus/src/homepage.rb:37]: 💯

**Review by davidsbailey — APPROVED** [review]
Super glad to see this go out with test coverage. ⚓ ship it!

File: dashboard/test/lib/homepage/homepage_test.rb
Comment (megcrenshaw) [reply dashboard/test/lib/homepage/homepage_test.rb]: Oh good call –– @kelbyhawn can I go ahead and update this? I hate shipping code that needs a refactor 😅

Comment (kelbyhawn) [reply dashboard/test/lib/homepage/homepage_test.rb]: @megcrenshaw yes, absolutely! How much time do you think this'll add before it can be merged? (Only asking due to the AI banner dependency 🙃)

Comment (megcrenshaw) [reply dashboard/test/lib/homepage/homepage_test.rb]: Give me like 15 minutes –– I have it moved with tests passing, but cleaning up a few things. Then it needs to go through drone ... so end of day your time

Comment (megcrenshaw) [reply dashboard/test/lib/homepage/homepage_test.rb]: Just pushed new stuff. @davidsbailey do you want to take another look? It's my first rodeo writing a pegasus unit test, but the test structure didn't change. Perhaps if there are requested refactors, I can tackle them in a separate PR

**Review by davidsbailey — APPROVED** [review]
thanks for moving these tests Meg! LGTM once drone passes.

I do have one question: which bucket do we fall into when viewing these pages in local development?  I don't think the answer should block this PR, but worth considering what we want the local experience to be.

**Comment by kelbyhawn** [comment]
@megcrenshaw feel free to merge this w/o me when it's ready (unless it's better practice that I do it?)

**Comment by megcrenshaw** [comment]
> @megcrenshaw feel free to merge this w/o me when it's ready (unless it's better practice that I do it?)

Ah, good call! It's ready. I'll merge 🥳

**Comment by megcrenshaw** [comment]
> thanks for moving these tests Meg! LGTM once drone passes.
> 
> I do have one question: which bucket do we fall into when viewing these pages in local development? I don't think the answer should block this PR, but worth considering what we want the local experience to be.

@davidsbailey great question. When Kelby and I tested this locally, the `country` comes back as `RD` (reserved). What do you see are potential options here? Elijah had some ideas here https://codedotorg.slack.com/archives/CFTFD6BPV/p1681839473513649?thread_ts=1681767673.939399&cid=CFTFD6BPV

**Comment by davidsbailey** [comment]
> > thanks for moving these tests Meg! LGTM once drone passes.
> > I do have one question: which bucket do we fall into when viewing these pages in local development? I don't think the answer should block this PR, but worth considering what we want the local experience to be.
> 
> davidsbailey great question. When Kelby and I tested this locally, the `country` comes back as `RD` (reserved). What do you see are potential options here? Elijah had some ideas here https://codedotorg.slack.com/archives/CFTFD6BPV/p1681839473513649?thread_ts=1681767673.939399&cid=CFTFD6BPV

my input would just be to pick what you want the behavior to be, and have your code provide that behavior when the country code is `RD`. so, if you want it to give the US behavior, then `location_unknown_or_usa = request.country.nil? || request.country.to_s.casecmp?('us') || .... ('RD')`

**Comment by davidsbailey** [comment]
it sounds like this is time-sensitive, so feel free to do in a follow-up PR.

**Comment by megcrenshaw** [comment]
@kelbyhawn do you want the local setup to behave like the US? If so, I'll create a followup PR to address that.

**Comment by kelbyhawn** [comment]
> @kelbyhawn do you want the local setup to behave like the US? If so, I'll create a followup PR to address that.

@megcrenshaw yes, that would be great thanks!

**Pull request merged**
